### ISSUE TITLE
Argo can't run on GKE Autopilot

### ISSUE THREAD

**Issue by simster7** [argoproj/argo-workflows#5434]
# Summary

What change needs making?

Argo can't run on [GKE Autopilot](https://cloud.google.com/kubernetes-engine/docs/concepts/autopilot-overview), it doesn't conform to [its requirements](https://cloud.google.com/kubernetes-engine/docs/concepts/autopilot-overview#limits). See if we can meet the requirements or add options to meet them.

(More details to follow)

# Use Cases

When would you use this?

GKE Autopilot seems to be popular among GCP users. We should see if we can run on it.

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by alexec** [comment]
which restrictions?

**Comment by Matts966** [comment]
@alexec 

Maybe because of `/var/run/docker.sock`

```
admission webhook "validation.gatekeeper.sh" deni
ed the request: [denied by autogke-no-write-mode-hostpath] hostPath volume docker-sock used in container wait
uses path /var/run/docker.sock which is not allowed in Autopilot. Allowed path prefixes for hostPath volumes a
re: ["/var/log/"]. Requesting user: <system:serviceaccount:argo:argo> and groups: <["system:serviceaccounts",
"system:serviceaccounts:argo", "system:authenticated"]>
```

**Comment by simster7** [comment]
@Matts966 that issue can actually be solved by running with the PNS executor, but even then you run into another issue. I can't remember what it was when I tested it, but I can take another look soon.

**Comment by bkanuka** [comment]
Even after changing to PNS there's an error regarding `SYS_PTRACE`:

> admission webhook "validation.gatekeeper.sh" denied the request: [denied by autogke-default-linux-capabilities] linux capability <{"SYS_PTRACE"}> on container <wait> not allowed; Autopilot only allows the capabilities: <["SETPCAP", "MKNOD", "AUDIT_WRITE", "CHOWN", "NET_RAW", "DAC_OVERRIDE", "FOWNER", "FSETID", "KILL", "SETGID", "SETUID", "NET_BIND_SERVICE", "SYS_CHROOT", "SETFCAP"]>. Requesting user: <system:serviceaccount:default:argo> and groups: <["system:serviceaccounts", "system:serviceaccounts:default", "system:authenticated"]>

**Comment by sky29** [comment]
Anyone has any workaround for this ? (It's kind of blocker for me)

**Comment by yuvaldrori** [comment]
I was able to get cron workflows to work by changing `containerRuntimeExecutor` from `docker` to `k8sapi` in the `workflow-controller-configmap`.

**Comment by alexec** [comment]
I believe you can run GKE using the Emissary executor in v3.1 (full featured) or the K8s API executor (>= v3.0, must use empty dir for artifacts). 

Unless anyone else knows better?

**Comment by Bobgy** [comment]
I verified emissary executor on GKE autopilot.
Works like a charm! : )

**Comment by lusid** [comment]
> Even after changing to PNS there's an error regarding `SYS_PTRACE`:
> 
> > admission webhook "validation.gatekeeper.sh" denied the request: [denied by autogke-default-linux-capabilities] linux capability <{"SYS_PTRACE"}> on container  not allowed; Autopilot only allows the capabilities: <["SETPCAP", "MKNOD", "AUDIT_WRITE", "CHOWN", "NET_RAW", "DAC_OVERRIDE", "FOWNER", "FSETID", "KILL", "SETGID", "SETUID", "NET_BIND_SERVICE", "SYS_CHROOT", "SETFCAP"]>. Requesting user: system:serviceaccount:default:argo and groups: <["system:serviceaccounts", "system:serviceaccounts:default", "system:authenticated"]>

Anyone know if this is resolved for PNS executor in GKE 1.21+ since SYS_PTRACE is supported?
https://cloud.google.com/kubernetes-engine/docs/concepts/autopilot-overview?authuser=1&_ga=2.36455523.-294381163.1588290990#linux_workload_limitations
